File: src/test/java/de/zalando/zally/rules/UseProblemJsonRuleTest.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Swagger;
import io.swagger.parser.SwaggerParser;
import org.junit.Test;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.assertTrue;

public class UseProblemJsonRuleTest {

    @Test
    public void responseUseProblemJsonRuleNoViolationsJson() throws IOException {
        UseProblemJsonRule rule = new UseProblemJsonRule();
        Swagger swagger = new SwaggerParser().read("/api_spp.json");
        List<Violation> violations = rule.validate(swagger);
        assertThat(violations.size()).isEqualTo(2);
        assertTrue(violations.stream().allMatch(v -> ViolationType.MUST.equals(v.getViolationType())));
        assertTrue(violations.stream().allMatch(v -> "Use Problem JSON".equals(v.getTitle())));
    }

    @Test
    public void responseUseProblemJsonRuleNoViolationYaml() throws IOException {
        UseProblemJsonRule rule = new UseProblemJsonRule();
        Swagger swagger = new SwaggerParser().read("/api_spa.yaml");
        assertThat(rule.validate(swagger)).isEmpty();
    }
}

File: server/src/test/java/de/zalando/zally/rules/UseProblemJsonRuleTest.kt
package de.zalando.zally.rules

import de.zalando.zally.getFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class UseProblemJsonRuleTest {

    @Test
    fun positiveCaseTinbox() {
        val swagger = getFixture("api_tinbox.yaml")
        assertThat(UseProblemJsonRule().validate(swagger)).isNull()
    }

    @Test
    fun negativeCaseSpp() {
        val swagger = getFixture("api_spp.json")
        val result = UseProblemJsonRule().validate(swagger)!!
        assertThat(result.paths).hasSameElementsAs(listOf(
                "/product-put-requests/{product_path} POST 400",
                "/product-put-requests/{product_path} POST 406",
                "/product-put-requests/{product_path} POST 413",
                "/product-put-requests/{product_path} POST 422",
                "/product-put-requests/{product_path} POST 429",
                "/products GET 429",
                "/products/{product_id} GET 418",
                "/products/{product_id} GET 429",
                "/products/{product_id} PATCH 400",
                "/products/{product_id} PATCH 429",
                "/products/{product_id} HEAD 404",
                "/products/{product_id} HEAD 429",
                "/products/{product_id}/children GET 418",
                "/products/{product_id}/children GET 429",
                "/products/{product_id}/updates/{update_id} GET 429",
                "/request-groups/{request_group_id}/summaries GET 429",
                "/request-groups/{request_group_id}/updates GET 429"
        ))
    }
}
